xilem_core/src/sequence.rs

```diff
@@ -1,363 +1,651 @@
-// Copyright 2023 the Xilem Authors
+// Copyright 2024 the Xilem Authors
 // SPDX-License-Identifier: Apache-2.0
 
-#[doc(hidden)]
-#[macro_export]
-macro_rules! impl_view_tuple {
-    ( $viewseq:ident, $elements_splice: ident, $pod:ty, $cx:ty, $changeflags:ty, $( $t:ident),* ; $( $i:tt ),* ) => {
-        impl<T, A, $( $t: $viewseq<T, A> ),* > $viewseq<T, A> for ( $( $t, )* ) {
-            type State = ( $( $t::State, )*);
-
-            fn build(&self, cx: &mut $cx, elements: &mut dyn $elements_splice) -> Self::State {
-                let b = ( $( self.$i.build(cx, elements), )* );
-                let state = ( $( b.$i, )*);
-                state
-            }
+//! Support for sequences of views with a shared element type.
 
-            fn rebuild(
-                &self,
-                cx: &mut $cx,
-                prev: &Self,
-                state: &mut Self::State,
-                els: &mut dyn $elements_splice,
-            ) -> ChangeFlags {
-                let mut changed = <$changeflags>::default();
-                $(
-                    let el_changed = self.$i.rebuild(cx, &prev.$i, &mut state.$i, els);
-                    changed |= el_changed;
-                )*
-                changed
-            }
+use alloc::vec::Vec;
 
-            fn message(
-                &self,
-                id_path: &[$crate::Id],
-                state: &mut Self::State,
-                message: Box<dyn std::any::Any>,
-                app_state: &mut T,
-            ) -> $crate::MessageResult<A> {
-                $crate::MessageResult::Stale(message)
-                $(
-                    .or(|message|{
-                        self.$i.message(id_path, &mut state.$i, message, app_state)
-                    })
-                )*
-            }
+use crate::{DynMessage, MessageResult, SuperElement, View, ViewElement, ViewId, ViewPathTracker};
 
-            fn count(&self, state: &Self::State) -> usize {
-                0
-                $(
-                    + self.$i.count(&state.$i)
-                )*
-            }
-        }
+/// An append only `Vec`.
+pub struct AppendVec<T> {
+    inner: Vec<T>,
+}
+
+impl<T> AppendVec<T> {
+    pub fn into_inner(self) -> Vec<T> {
+        self.inner
+    }
+    pub fn push(&mut self, item: T) {
+        self.inner.push(item);
+    }
+    pub fn drain(&mut self) -> impl Iterator<Item = T> + '_ {
```

Comment:
```suggestion
    pub fn drain(&mut self) -> Drain<'_, T> {
```

Or should this hide the underlying type/implementation details?
With the concrete type it e.g. also has the properties `impl ExactSizeIterator` and `impl DoubleEndedIterator`.